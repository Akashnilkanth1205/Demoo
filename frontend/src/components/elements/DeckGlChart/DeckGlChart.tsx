/**
 * @license
 * Copyright 2018-2020 Streamlit Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { ReactElement, useEffect, useState } from "react"
import Immutable from "immutable"
import { StaticMap } from "react-map-gl"
import withFullScreenWrapper from "hocs/withFullScreenWrapper"
import withMapboxToken from "hocs/withMapboxToken/withMapboxToken"
import "mapbox-gl/dist/mapbox-gl.css"
import "./DeckGlChart.scss"
import { buildLayer, getStyleUrl } from "./DeckGlChartUtil"
import DeckGL from "deck.gl"

export interface DeckGlChartProps {
  width: number
  mapboxToken: string
  element: Immutable.Map<string, any>
  height: number | undefined
}

interface ViewState {
  width: number
  height: number
  longitude: number
  latitude: number
  pitch: number
  bearing: number
  zoom: number
}

interface ViewPort extends ViewState {
  mapStyle: string
}

export function DeckGlChart({
  element,
  height = 500,
  mapboxToken,
  width,
}: DeckGlChartProps): ReactElement {
  const [initialized, setInitialized] = useState(false)

  useEffect(() => {
    // HACK: Load layers a little after loading the map, to hack around a bug
    // where HexagonLayers were not drawing on first load but did load when the
    // script got re-executed.
    setInitialized(true)
  }, [])

  const getViewport = (): ViewPort => {
    const specStr = element.get("spec")
    const spec = specStr ? JSON.parse(specStr) : {}

    return spec.viewport || {}
  }

  const generateViewState = (viewPort: ViewPort): ViewState => {
    const useContainerWidth = element.get("useContainerWidth")

    return {
      width: !viewPort.width || useContainerWidth ? width : viewPort.width,
      height: viewPort.height || height,
      longitude: viewPort.longitude || 0,
      latitude: viewPort.latitude || 0,
      pitch: viewPort.pitch || 0,
      bearing: viewPort.bearing || 0,
      zoom: viewPort.zoom || 1,
    } as ViewState
  }

  const viewPort = getViewport()

  const initialViewState = generateViewState(viewPort)
  const mapStyle = getStyleUrl(viewPort.mapStyle)
  const buildLayers = (): any[] => {
    const layers = element.get("layers")
    return layers.map((layer: any) => buildLayer(layer)).toArray()
  }
  return (
    <div
      className="deckglchart stDeckGlChart"
      style={{
        height: initialViewState.height,
        width: initialViewState.width,
      }}
    >
      <DeckGL
        initialViewState={initialViewState}
        height={initialViewState.height}
        width={initialViewState.width}
        controller
        layers={initialized ? buildLayers() : []}
      >
        <StaticMap
          height={initialViewState.height}
          width={initialViewState.width}
          mapStyle={mapStyle}
          mapboxApiAccessToken={mapboxToken}
        />
      </DeckGL>
    </div>
  )
}

export default withMapboxToken("st.deck_gl_chart")(
  withFullScreenWrapper(DeckGlChart)
)
