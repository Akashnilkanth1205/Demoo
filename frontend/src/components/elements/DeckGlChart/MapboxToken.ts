/**
 * @license
 * Copyright 2018-2020 Streamlit Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { SessionInfo } from "lib/SessionInfo"

export const TOKENS_URL = "https://streamlit.io/tokens.json"

/**
 * Exposes a singleton MapboxToken. If the user has specified
 * their own
 */
export class MapboxToken {
  private static token?: string

  public static async get(): Promise<string> {
    if (MapboxToken.token == null) {
      if (SessionInfo.current.userMapboxToken !== "") {
        MapboxToken.token = SessionInfo.current.userMapboxToken
      } else {
        MapboxToken.token = await this.fetchToken(TOKENS_URL, "mapbox")
      }
    }

    return MapboxToken.token
  }

  private static async fetchToken(
    url: string,
    tokenName: string
  ): Promise<string> {
    const rsp = await fetch(url)
    if (!rsp.ok) {
      throw new Error(`${url}: Bad status ${rsp.status}`)
    }

    const json = await rsp.json()
    const token = json[tokenName]
    if (token == null || token === "") {
      throw new Error(`${url}: Missing token "${tokenName}"`)
    }

    return token
  }
}
