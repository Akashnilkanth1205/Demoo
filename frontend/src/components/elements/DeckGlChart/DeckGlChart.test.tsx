/**
 * @license
 * Copyright 2018-2020 Streamlit Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from "react"
import DeckGL from "deck.gl"
import { mount } from "enzyme"
import { fromJS } from "immutable"
import { StaticMap } from "react-map-gl"

import mockDeckGl from "./mock"
import { DeckGlChart, DeckGlChartProps } from "./DeckGlChart"

const getProps = (
  elementProps: Record<string, unknown> = {}
): DeckGlChartProps => ({
  element: fromJS({
    ...mockDeckGl,
    ...elementProps,
  }),
  width: 698,
  height: 500,
  mapboxToken: "mapboxToken",
})

describe("DeckGlChart Element", () => {
  const props = getProps()
  const wrapper = mount(<DeckGlChart {...props} />)

  it("renders without crashing", () => {
    expect(wrapper).toBeDefined()
  })

  it("should have proper shape", () => {
    const wrappingDiv = wrapper.find("div").first()

    expect(wrappingDiv.prop("className")).toContain("stDeckGlChart")
    expect(wrappingDiv.prop("style")).toStrictEqual({
      height: props.height,
      width: props.width,
    })
    expect(wrapper.find(DeckGL).prop("height")).toBe(props.height)
    expect(wrapper.find(DeckGL).prop("width")).toBe(props.width)
  })

  it("should have a DeckGL", () => {
    expect(wrapper.find(DeckGL).prop("initialViewState")).toMatchSnapshot(
      "initialViewState"
    )

    expect(wrapper.find(DeckGL).props()).toHaveProperty("controller")
    expect(wrapper.find(DeckGL).prop("layers")).toMatchSnapshot()
    // to check why this fails
    /* expect(wrapper.find(StaticMap).prop("mapboxApiAccessToken")).toBe(
      props.mapboxToken
    )
    */
  })
})
