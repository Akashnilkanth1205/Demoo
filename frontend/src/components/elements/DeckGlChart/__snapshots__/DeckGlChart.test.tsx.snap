// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeckGlChart Element should have a DeckGL 1`] = `Array []`;

exports[`DeckGlChart Element should have a DeckGL: initialViewState 1`] = `
Object {
  "bearing": 0,
  "height": 500,
  "latitude": 37.76,
  "longitude": -122.4,
  "pitch": 50,
  "width": 698,
  "zoom": 11,
}
`;
